Added tests for LongitudinalProfileFromData and fix for issue #185

em-archer opened the pull request:
Editing LongitudinalProfileFromData to allow for monotonically decreasing spectral data (fix for #185) and adding relevant tests to test_laser_profiles. Currently no tests exist for LongitudinalProfileFromData, so adding tests for monotonically increasing/decreasing spectral data. Test for temporal data could be added in future.

Commit:
Added tests for LongitudinalProfileFromData

Adding tests to allow for longitudinal profile from spectral data, whether monotonically increasing or decreasing

Commit:
Monotonically decreasing spectral data in LongitudinalProfileFromData

Editing LongitudinalProfileFromData to allow for monotonically decreasing spectral data

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Updated tests

Updated tests for monotonically increasing or decreasing specta

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Minor correction

Commit:
em-archer

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Minor correction

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Syntax

Commit:
Syntax conflict

Wavelength and wavelength axis given different names

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
em-archer

Commit:
em-archer

Commit:
Should now include phase

Analytic ft of GaussianLongitudinalProfile needs checking

Commit:
Final commit

Analytic spectrum checked in separate script

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Fixing flipping of arrays

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
em-archer

Commit:
Assertions also checked with separate notebook

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Increase time resolution

Commit:
Modified to also allow spectral data to be specified with respect to omega

When data["datatype"] = "spectral", data["axis"] can either be wavelength lambda in m or angular frequency omega in s^-1. The user doesn't need to specify, it will automatically detect and convert to wavelength if the latter is passed.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

em-archer commented:
Also added small update to also accept angular spectrum data (units of s^-1) instead of wavelength data

MaxThevenet commented on tests/test_laser_profiles.py:
```diff
@@ -234,6 +238,89 @@ def test_longitudinal_profiles():
 
+    # LongitudinalProfileFromData
+    print("LongitudinalProfileFromData")
+    data = {}  # Generate spectral data assuming analytic ft of GaussianLongitudinalProfile
```
Is this correct? Just to make it more explicit.
```suggestion
    data = {}  # Generate spectral data assuming analytic Fourier transform of GaussianLongitudinalProfile
```

MaxThevenet reviewed (commented):
Thanks for this PR! Really great to have more tests, and the option to initialise from file while passing the angular frequency instead of wavelength will be useful, in particular for @Paaaaarth in PR #296. Besides a minor suggestion in the text, I have two comments:
1. Determining whether it is wavelength or frequency depending on the value is IMO not very straightforward. Would it be possible to make the user provide specific arguments? The user could provide `datatype = spectral` as before, but there could be one additional argument `axis_is_wavelength` with default `True`, and if set to `False` the axis is supposed to be frequency. I think this parameter would appear clearly in the doc and should make it clear to the users.
2. When converting from axis = frequency to wavelength, shouldn't we also modify the spectrum by multiplying with $2\pi c / \lambda^2$ or so? Not sure this is done in the current reader, could you check this?

em-archer committed:
Merge branch 'LASY-org:development' into development

Commit:
Added requirement for user to specify axis type

User must now specify axis type (wavelength by default). Also added proper conversion for spectral intensity if axis is omega.

em-archer committed:
Merge branch 'LASY-org:development' into development

Commit:
Merge branch 'development' of https://github.com/em-archer/lasy into development

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MaxThevenet commented on lasy/profiles/longitudinal/longitudinal_profile_from_data.py:
```diff
@@ -60,16 +63,27 @@ class LongitudinalProfileFromData(LongitudinalProfile):
-            ), 'data["axis"] must be in monotonously increasing order.'
-            spectral_intensity = data["intensity"]
+            if "axis_is_wavelength" not in data:  # Set to wavelength data by default
+                data["axis_is_wavelength"] = True
```
I think this will modify the actual `data` object passed by the user: if the user defines an actual `data` variable without `"axis_is_wavelength"` and passes it to this function, it will have this attribute defined after this call.
```py
data = ... # "axis_is_wavelength" not in data
p = LongitudinalProfileFromData(data, ...)
# now "axis_is_wavelength" is in data
```
In most cases, this will have absolutely no importance. But I think it is still safer to avoid modifying it. The solution is simple, define a new variable, and use this one instead, e.g.
```py
if "axis_is_wavelength" not in data:
    axis_is_wavelength = True
else:
    axis_is_wavelength = data["axis_is_wavelength"]
```

em-archer committed:
Merge branch 'LASY-org:development' into development

MaxThevenet commented on lasy/profiles/longitudinal/longitudinal_profile_from_data.py:
```diff
@@ -25,12 +25,15 @@ class LongitudinalProfileFromData(LongitudinalProfile):
         datatype : string
             The domain in which the data has been passed. Options
-            are 'spectral' and 'temporal'
+            are 'spectral' and 'temporal'. Spectral data may either be,
+            a) Wavelength dependent (lambda): data["axis_is_wavelength"] = True or
+            b) Angular freuqency dependent (omega): data["axis_is_wavelength"] = False
```
Considering suggestion below.
```suggestion
            are 'spectral' and 'temporal'.
```

MaxThevenet commented on lasy/profiles/longitudinal/longitudinal_profile_from_data.py:
```diff
@@ -25,12 +25,15 @@ class LongitudinalProfileFromData(LongitudinalProfile):
+            are 'spectral' and 'temporal'. Spectral data may either be,
+            a) Wavelength dependent (lambda): data["axis_is_wavelength"] = True or
+            b) Angular freuqency dependent (omega): data["axis_is_wavelength"] = False
 
```
This option probably warrants a separate entry, here's a suggestion for this.
```suggestion

        axis_is_wavelength : boolean (optional, default True)
            If True, the axis represents wavelength in [m] (SI).
            If False, it represents frequency in [1/s] (SI).

```

MaxThevenet commented on lasy/profiles/longitudinal/longitudinal_profile_from_data.py:
```diff
@@ -25,12 +25,15 @@ class LongitudinalProfileFromData(LongitudinalProfile):
-            When datatype is 'spectral' axis is wavelength in meters.
+            The array must be monotonically increasing or decreasing.
+            When datatype is 'spectral' axis is wavelength in meters OR
+            angular frequency in 1/seconds.
```
```suggestion
            angular frequency in 1/seconds.
```

MaxThevenet reviewed (commented):
Thanks! See a few comments below, I think that's the last round.

em-archer committed:
Update lasy/profiles/longitudinal/longitudinal_profile_from_data.py

Co-authored-by: Maxence Thévenet <maxence.thevenet@desy.de>

em-archer committed:
Update lasy/profiles/longitudinal/longitudinal_profile_from_data.py

Co-authored-by: Maxence Thévenet <maxence.thevenet@desy.de>

MaxThevenet commented:
Superseded by #309.
